src/executor/graph_executor.cc

```diff
@@ -1167,9 +1234,37 @@ Graph GraphExecutor::InitGraph(nnvm::Symbol symbol,
                                const std::vector<Context>& in_arg_ctxes,
                                const std::vector<Context>& arg_grad_ctxes,
                                const std::vector<Context>& aux_state_ctxes,
+                               nnvm::ShapeVector &arg_shapes,
+                               nnvm::DTypeVector &arg_dtypes,
+                               StorageTypeVector &arg_stypes,
                                const std::vector<OpReqType>& grad_req_types) {
+  
+  // parition given symbol graph using shape & types.
+  nnvm::Graph gp;
+  gp.outputs = symbol.outputs;
+  auto num_forward_outputs = symbol.outputs.size();
+  auto num_forward_inputs = symbol.ListInputs(nnvm::Symbol::kAll).size();
+  gp = AssignContext(gp, default_ctx, ctx_map,
+                    in_arg_ctxes,
+                    arg_grad_ctxes,
+                    aux_state_ctxes,
+                    grad_req_types,
+                    num_forward_inputs,
+                    num_forward_outputs);
+
+  // infer & add types/shape attrs to graph before calling partition pass
+  Infer(gp, num_forward_inputs, arg_shapes, arg_dtypes, arg_stypes);
+
+  // partition pass with default subgraph property
+  mxnet::op::SubgraphPropertyPtr property =
+      std::make_shared<mxnet::op::ShapeTypesSubgraphProperty>();
```

Comment:
we need to make it general for all kinds of backends.
one option is to use a hashtable whose key is the backend name, the value is the SubgraphPropertyPtr for the particular backend.
We can use the environment variable to check which backend should be used.

Reply:
will look over this more, we need to also consider if we can use context for this ?